src/sinks/util/uri.rs

```diff
@@ -1,18 +1,38 @@
-use http::Uri;
+use crate::http::Auth;
+use http::uri::{Authority, Uri};
+use percent_encoding::percent_decode_str;
 use serde::{
     de::{Error, Visitor},
     Deserialize, Deserializer, Serialize, Serializer,
 };
 use std::fmt;
+use std::str::FromStr;
 
 /// A wrapper for `http::Uri` that implements the serde traits.
+/// Authorization credentials, if exist, will be removed from the URI and stored in `auth`.
+/// For example: "http://user:password@example.com".
 #[derive(Default, Debug, Clone)]
-pub struct UriSerde(Uri);
+pub struct UriSerde {
+    pub uri: Uri,
+    pub auth: Option<Auth>,
+}
+
+impl UriSerde {
+    /// Used to combine existing authorization credentials with credentials in this URI.
+    /// If both is `Some`, return an error. Otherwise, choose one of them.
+    pub fn merge_auth_config(&self, auth: &mut Option<Auth>) -> crate::Result<()> {
+        if auth.is_some() && self.auth.is_some() {
+            Err("Two authorization credentials was provided.".into())
+        } else {
+            *auth = auth.take().or_else(|| self.auth.clone());
```

Comment:
I think this function makes more sense on `Auth` since it mutates it.

Reply:
I moved it to `fn merge_auth_config(&mut Option<Auth>, &Option<Auth>)` inside `impl Auth`. Unfortunately we cannot use an associated method because it need `self` to be `Option<Auth>`.

Reply:
Instead of requiring `self` to be `Option<Auth>` could we just have callers use `maybe_auth.map(...)`?

Reply:
Never mind, I see the function signature changed to merging two `Auth`s instead.